horace_core/utilities/split_vector_fixed_sum.m

```diff
@@ -0,0 +1,217 @@
+function [chunks, idxs] = split_vector_fixed_sum(numeric_vector, chunk_sum, cumulative_sum)
+%SPLIT_VECTOR_FIXED_SUM Split the given vector into sub-vectors such that each
+% sub-vector has sum 'chunk_sum'. If chunk sum does not divide the sum of the
+% input vector, the final chunk's sum will be the remainder.
+%
+% Values in the input vector may be split between chunks.
+%
+% Input:
+% ------
+% numeric_vector   A vector of numeric, non-negative, values.
+% chunk_sum        A positive value specifying the sum for each sub-vector.
+% cumulative_sum   The cumulative sum of 'numeric_vector'. (optional)
+%
+% Output:
+% -------
+% chunks   Cell array of vectors.
+% idxs     The indices at which the input vector was "split". Has size
+%          [2, n], where n is numel(chunks). Each idxs(:, chunk_num) is the
+%          lower and upper index into 'numeric_vector' of chunks{chunk_num}.
+%
+% Example:
+% --------
+% >> numeric_vector = [3, 2, 6, 0, 5, 3, 1, 0, 1, 24, 4, 2, 3, 0];
+% >> chunk_sum = 10;
+% >> [chunks, idxs] = split_vector_max_sum(numeric_vector, chunk_sum)
+%   chunks =
+%       { [3, 2, 5], [1, 0, 5, 3, 1, 0], [1, 9], [10], [5, 4, 1], [1, 3, 0] }
+%   idxs =
+%       1     4     9    10    10    12
+%       4     8    10    10    12    14
+%
+% >> numeric_vector = [24];
+% >> chunk_sum = 10;
+%   chunks =
+%       { [10], [10], [4] }
+%   idxs =
+%       1
+%       1
+%
+if isempty(numeric_vector)
+    chunks = {};
+    idxs = zeros(2, 0);
+    return;
+end
+
+validateattributes(numeric_vector, {'numeric'}, {'vector', 'nonnegative'});
+validateattributes(chunk_sum, {'numeric'}, {'scalar', 'positive'});
+
+if ~exist('cumulative_sum', 'var')
+    cumulative_sum = cumsum(numeric_vector);
+end
+vector_sum = cumulative_sum(end);
```

Comment:
Is there an easy check you can do to verify cumulative_sum is reasonable? Or do we rely on good faith?

Reply:
e.g. 
```matlab
split_vector_fixed_sum([1,2,4,4], 3, [1])
```

Reply:
The intention is that, generally, you do not use the argument. But there were some cases where the cumulative sum was used elsewhere in the calling function, which made it advantages to be able to just pass in here, rather than calculate it again.

Reply:
@oerc0122 raises a good point.

We could easily add a check that the final value in cumsum is equal to sum - at a performance cost. How critical is the performance of this function? At least part of the reason for taking cumsum as an argument is to stop its unnecessary recalculation.

An alternative is to add a note to the function header that cumsum argument isn't validated and the behaviour if it's not correct is undetermined.

Reply:
I ran some benchmarks for cumsum/summing a large array:
```matlab
>> x = rand(1, 1e8);
>> median(benchmark_function(@() sum(x), 10, 10))

ans =

    0.0249

>> median(benchmark_function(@() cumsum(x), 10, 10))

ans =

    0.2215
```

Looks like it's very cheap to get the sum, but that calls into question the value of providing the `cumulative_sum` argument anyway.

The callers all use `cumulative_sum(end)` to verify that the length of a pixel array matches the sum of the elements in an `npix` array. Perhaps the cleanest solution is to remove the `cumulative_sum` argument from `split_vector_` and just use a `sum` to perform validation in the calling functions (e.g. `do_binary_op_sigvar_`).